### PR TITLE
Settings user only load avatar if available

### LINKED ISSUE #20644: User list unresponsible
> :-1:  Pretty slow for me, mainly with default system avatars which takes ages to load. Takes over a minute, with LDAP running on the same machine.
>
before2

From https://github.com/owncloud/core/pull/20108#issuecomment-151866829

cc @butonic @blizzz

### PR DESCRIPTION
**Pull request by rullzer**
Fixes https://github.com/owncloud/core/issues/20644

While generating the list of users check directly if the avatar is available. If it is not then the user does not even have to try to load it.

The assumption here is that if an installation has a large number of users it is likely that a lot of them do not have avatars. This then reduces significantly the number of requests to the server (which should also limit the load on the server).

So basically we trade of time vs time. Generating the user list will take slightly longer but this will result most likely in significant less requests.

CC: @MorrisJobke @PVince81 @blizzz

### CODE REVIEW

**Comment by rullzer** [comment]
To test

* create an instance with N users where M of them have an avatar.
* Go to the user page
* Observer the number of avatars requested

On master there will be N
On this branch M

File: settings/controller/userscontroller.php [hunk 1]
```diff
@@ -179,6 +184,7 @@ private function formatUserForIndex(IUser $user, array $userGroups = null) {
 			'backend' => $user->getBackendClassName(),
 			'email' => $displayName,
 			'isRestoreDisabled' => !$restorePossible,
+			'isAvatarAvailable' => $this->avatarManager->getAvatar($user->getUID())->exists()
```
Comment (PVince81) [inline settings/controller/userscontroller.php]: How expensive is `getAvatar()` ? You might want to remove this line whenever avatars are disabled system-wide, to save another call.

**Comment by PVince81** [comment]
Nice idea!

File: settings/controller/userscontroller.php (hunk 1 above)
Comment (rullzer) [inline settings/controller/userscontroller.php]: Uhm basically a fs query. But otherwise we have to do it anyway because then the user pages requests the avatar and we also check it ;)

Ah yes if avatars are disabled we can set it do false by default
lets fix that

Comment (rullzer) [inline settings/controller/userscontroller.php]: Fixed

**Comment by blizzz** [comment]
I like the idea and the stuff works so :+1:  however, i do not feel any significant difference with plentiful of users (from LDAP via VPN), looks more or less the same. Cannot say it fixes the responsiveness as described in #20644. Time to deal with requests is decreased, but it still takes a significant amount of time to render the whole stuff. I can add numbers later.

**Comment by blizzz** [comment]
Master without avatars:

a-requests-master-no-avatars1

Master with avatars:

a-requests-master

This branch with  avatars:

a-requests-fixbranch

My test instance were two LDAPs, one connected via VPN (which should not happen in a practical setup, it's a bad scenario), i.e. i think 500+couple local users. Only very few with Avatars. With master, if i remember correctly, having an avatar was better then having the placeholder.

So in fact it's already a huge improvement. We are still fetching too many users, or our rendering code is too bad, or both.

**Comment by MorrisJobke** [comment]
I just tested this with a zombie army that doesn't have avatars -> load is down to 1 second (from 10 seconds) :+1: with avatars it stays nearly the same.

**Comment by rullzer** [comment]
@blizzz well we can't really help the rendering.since that is just done by your browser.

**Comment by blizzz** [comment]
> @blizzz well we can't really help the rendering.since that is just done by your browser.

Well, if our client JS can be optimized (did not check whether there is potential) or less entries need to be loaded.
